File: fp-in-scala/src/main/scala/fpinscala/answers/datastructures/Tree.scala
package fpinscala.answers.datastructures

enum Tree[+A]:
  case Leaf(value: A)
  case Branch(left: Tree[A], right: Tree[A])

  def size: Int = this match
    case Leaf(_)      => 1
    case Branch(l, r) => 1 + l.size + r.size

  // Exercise 3.26: ツリーの深さを計算するメソッド `depth` を定義せよ。深さは、ルートから最も遠いリーフまでのパスの長さである。

  def depth: Int = this match
    case Leaf(_)      => 0
    case Branch(l, r) => 1 + (l.depth.max(r.depth))

  // Exercise 3.27: ツリーの各リーフに関数 `f` を適用するメソッド `map` を定義せよ。

  def map[B](f: A => B): Tree[B] = this match
    case Leaf(a)      => Leaf(f(a))
    case Branch(l, r) => Branch(l.map(f), r.map(f))

  // Exercise 3.28-1: ツリーのリーフの値を変換する関数 `f` とブランチの左右の値をまとめる関数 `g` を受け取ってツリーを畳み込むメソッド `fold` を定義せよ。
  // また、 `fold` を用いて `size` 、 `depth` 、 `map` を定義せよ。

  /*
  Like `foldRight` for lists, `fold` receives a "handler" for each of the data constructors of the type, and recursively
  accumulates some value using these handlers. As with `foldRight`, `t.fold(Leaf(_))(Branch(_,_)) == t`, and we can use
  this function to implement just about any recursive function that would otherwise be defined by pattern matching.
   */
  def fold[B](f: A => B, g: (B, B) => B): B = this match
    case Leaf(a)      => f(a)
    case Branch(l, r) => g(l.fold(f, g), r.fold(f, g))

  def sizeViaFold: Int =
    fold(a => 1, 1 + _ + _)

  def depthViaFold: Int =
    fold(a => 0, (d1, d2) => 1 + (d1 max d2))

  def mapViaFold[B](f: A => B): Tree[B] =
    fold(a => Leaf(f(a)), Branch(_, _))

object Tree:

  def size[A](t: Tree[A]): Int = t match
    case Leaf(_)      => 1
    case Branch(l, r) => 1 + size(l) + size(r)

  extension (t: Tree[Int])
    def firstPositive: Int = t match
      case Leaf(i)      => i
      case Branch(l, r) =>
        val lpos = l.firstPositive
        if lpos > 0 then lpos else r.firstPositive

  // Exercise 3.25: ツリーのリーフの最大値を計算する拡張メソッド `maximum` を定義せよ。

  /*
  We're using the method `max` that exists on all `Int` values rather than an explicit `if` expression.

  Note how similar the implementation is to `size`. We'll abstract out the common pattern in a later exercise.
   */
  extension (t: Tree[Int])
    def maximum: Int = t match
      case Leaf(n)      => n
      case Branch(l, r) => l.maximum.max(r.maximum)

  // Exercise 3.28-2: `fold` を用いて `maximum` を定義せよ。

  extension (t: Tree[Int])
    def maximumViaFold: Int =
      t.fold(a => a, _ max _)

File: fp-in-ocaml/answers/datastructures/tree.ml
(** List.t と同様 *)
type +'a t = Leaf of 'a | Branch of 'a t * 'a t

let rec size = function Leaf _ -> 1 | Branch (l, r) -> 1 + size l + size r

(** Exercise 3.25: ツリーのリーフの最大値を計算する関数[maximum]を定義せよ。 *)
let rec maximum : int t -> int = function
  | Leaf v -> v
  | Branch (l, r) -> max (maximum l) (maximum r)

(** Exercise 3.26: ツリーの深さを計算する関数[depth]を定義せよ。 深さは、ルートから最も遠いリーフまでのパスの長さである。 *)
let rec depth : 'a t -> int = function
  | Leaf _ -> 0
  | Branch (l, r) -> 1 + max (depth l) (depth r)

(** Exercise 3.27: ツリーの各リーフに関数[f]を適用する関数[map]を定義せよ。 *)
let rec map (f : 'a -> 'b) : 'a t -> 'b t = function
  | Leaf v -> Leaf (f v)
  | Branch (l, r) -> Branch (map f l, map f r)

(** Exercise 3.28-1: ツリーのリーフの値を変換する関数[f]とブランチの左右の値をまとめる関数[g]を受け取って
    ツリーを畳み込む関数[fold]を定義せよ。 また、[fold]を用いて[size_via_fold], [depth_via_fold],
    [map_via_fold]を定義せよ。 *)
let rec fold (f : 'a -> 'b) (g : 'b * 'b -> 'b) : 'a t -> 'b = function
  | Leaf v -> f v
  | Branch (l, r) -> g (fold f g l, fold f g r)

let size_via_fold t = fold (Fun.const 1) (fun (l, r) -> 1 + l + r) t
let depth_via_fold t = fold (Fun.const 0) (fun (l, r) -> 1 + max l r) t

let map_via_fold f t =
  fold (fun v -> Leaf (f v)) (fun (l, r) -> Branch (l, r)) t

(** Exercise 3.28-2: [fold]を用いて[maximum]を定義せよ。 *)
let maximum_via_fold t = fold Fun.id (fun (l, r) -> max l r) t
